TLM/CSUtil.Commons/VectorUtil.cs

```diff
@@ -1,22 +1,27 @@
-﻿using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
-using UnityEngine;
+﻿namespace CSUtil.Commons {
+    using UnityEngine;
 
-namespace CSUtil.Commons {
-	public static class VectorUtil {
-		public static void ClampRectToScreen(ref Rect rect, Vector2 resolution) {
-			Log._Debug($"ClampPosToScreen([{rect.x}, {rect.y}, {rect.xMax}, {rect.yMax}], [{resolution.x}, {resolution.y}]) called");
-			if (rect.x < 0)
-				rect.x = 0;
-			if (rect.y < 0)
-				rect.y = 0;
-			if (rect.xMax >= resolution.x)
-				rect.x = resolution.x - rect.width;
-			if (rect.yMax >= resolution.y)
-				rect.y = resolution.y - rect.height;
-			Log._Debug($"ClampPosToScreen() -> [{rect.x}, {rect.y}, {rect.xMax}, {rect.yMax}]");
-		}
-	}
-}
+    public static class VectorUtil {
+        public static void ClampRectToScreen(ref Rect rect, Vector2 resolution) {
```

Comment:
could we not use a `ref` for the `resolution` vect? seems pointless to make a copy of it

Reply:
wait isn't `ref Type` the opposite to copying, i.e. passing the reference to the data stored somewhere else?

Reply:
yes, it is, with ref keyword we will get here reference to struct, not copy of

Reply:
so ref = less memalloc and gc ?

Reply:
Yes

Reply:
yes exactly, same as C/C++ `&variable` borrowing a writable reference but not moving the whole data

Reply:
you can use `in` keyword for resolution. It makes an argument read-only reference. (C#7.2 feature)

Reply:
We should probably start using `in` more often then (for stuff that lives 100% in the main thread).

Reply:
>for stuff that lives 100% in the main thread

So basically everything besides pathfinding (CustomPathFind2 class)

Reply:
Difference between `in` and `ref` is likely just compile-time check to ensure nothing in the method tries altering the value?

Reply:
nope it's still visible when you use e.g. dnSpy on compiled dlls
![inref](https://user-images.githubusercontent.com/19638970/61336085-3f92e980-a830-11e9-8e5b-a30d659789ff.jpg)

Reply:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/in-parameter-modifier

Reply:
I guess that also means it works with .net 3.5?

Reply:
I've changed that, compiled and no issues in-game yet.

Reply:
I imagine it's used when dragging the TM:PE button / menu?

Reply:
Right, but code can be optimized there.

Reply:
Is there a codemaid (or whatever it's called) setting to suggest where `in` and `ref` can be used?

Also, when using `in` (and `ref`?) it would be good to include those keywords at the call site too to make it obvious what's happening without needing to go look at the called method. Maybe a codemaid setting to encourage that too?

Reply:
There is an extension for VS and Rider which shows possible gc and memalloc's
https://github.com/microsoft/RoslynClrHeapAllocationAnalyzer
https://plugins.jetbrains.com/plugin/9223-heap-allocations-viewer

Reply:
Installed :) Would be nice addition to CI if we can add it in?

I've made new [Dev tools](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/wiki/Dev-Tools) page in the wiki to keep track of stuff like this.

If you know more tools like that just edit wiki or dump url's in to #developer chat (@ me so I can find more easily)

Reply:
For CI, what would you do about the fact that some data is boxed, allocated or copied? An excessive copy might be a minor warning, but not a build-breaking warning.

Reply:
Please don't use `in` "all the way". It's a very dangerous feature and will almost surely kill the performance. See the discussion in #349 for details.